feat: add support for `compileFunction` allowing us to avoid the module wrapper

SimenB opened the pull request:
<!-- Thanks for submitting a pull request! Please provide enough information so that others can review your pull request. The two fields below are mandatory. -->

<!-- Please remember to update CHANGELOG.md in the root of the project if you have not done so. -->

## Summary

This is essentially https://github.com/nodejs/node/pull/21573. Doing this should make it easier for sourcemaps to do their thing since we don't inject arbitrary code around the code.

I made this into an option as `vm.compileFunction` was added `node@10.10`. Happy to not have an option and rather use feature detection? Also note that JSDOM does not support it, but I'll open up a PR shortly.

EDIT: https://github.com/jsdom/jsdom/pull/2731

Mildly related to #8596 as it won't need the `moduleWrapperLength` thing anymore.

Docs: https://nodejs.org/api/vm.html#vm_vm_compilefunction_code_params_options

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->

## Test plan

I haven't added new tests yet, but existing tests should pass, na drunning locally with `--no-module-wrapper` seems to work correctly. If CI is happy, I'll flip the default and see what it says - hopefully the only failing tests at that point will be the ones for config

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

SimenB commented on CHANGELOG.md:
```diff
@@ -25,11 +26,13 @@
 - `[jest-snapshot]` Improve colors when snapshots are updatable ([#9132](https://github.com/facebook/jest/pull/9132))
 - `[jest-snapshot]` Ignore indentation for most serialized objects ([#9203](https://github.com/facebook/jest/pull/9203))
 - `[jest-transform]` Create `createTranspilingRequire` function for easy transpiling modules ([#9194](https://github.com/facebook/jest/pull/9194))
+- `[jest-transform]` [**BREAKING**] Do not automatically wrap functions in a module wrapper
```
extracted this breaking change into #9253

SimenB commented on packages/jest-environment-jsdom/src/index.ts:
```diff
@@ -130,6 +130,14 @@ class JSDOMEnvironment implements JestEnvironment {
+  compileFunction(code: string, params: Array<string>, filename: string) {
+    if (this.dom) {
+      // @ts-ignore
+      return this.dom.compileFunction(code, params, {filename}) as any;
```
`@ts-ignore` due to https://github.com/jsdom/jsdom/pull/2731 not being merged.

SimenB commented on packages/jest-environment-node/src/index.ts:
```diff
@@ -110,6 +116,22 @@ class NodeEnvironment implements JestEnvironment {
+  compileFunction(code: string, params: Array<string>, filename: string) {
+    if (!compileFunction) {
+      throw new TypeError(
+        "The version of Node you're using does not support compileFunction, please upgrade",
```
dunno if we want a more specific error message?

thymikee replied:
I think taking the minimum version from the Nod docs would be helpful here

SimenB commented on packages/jest-environment-node/src/index.ts:
```diff
@@ -110,6 +116,22 @@ class NodeEnvironment implements JestEnvironment {
+      return compileFunction(code, params, {
+        filename,
+        parsingContext: this.context,
+      }) as any;
```
it returns `Function`

thymikee commented on packages/jest-config/src/Defaults.ts:
```diff
@@ -38,6 +38,7 @@ const defaultOptions: Config.DefaultOptions = {
   moduleFileExtensions: ['js', 'json', 'jsx', 'ts', 'tsx', 'node'],
   moduleNameMapper: {},
   modulePathIgnorePatterns: [],
+  moduleWrapper: true,
```
If we can avoid config and feature-detect, let's do it please

SimenB replied:
main reason I added it was so people could opt out if it was buggy. It should be safe to just enable though, as long as all of our own tests still pass

thymikee commented on packages/jest-transform/src/ScriptTransformer.ts:
```diff
@@ -382,6 +404,7 @@ export default class ScriptTransformer {
           displayErrors: true,
           filename: isCoreModule ? 'jest-nodejs-core-' + filename : filename,
         }),
+        scriptContent: wrappedCode,
```
I hope this won't increase memory footprint too much (thinking about some big scripts like TS main file)

SimenB replied:
we can make sure to not return the `Script` if we haven't wrapped anything as it'll be broken aynway

SimenB replied:
I've done that, although we'll need access to the pre-transformed source as well for v8 coverage. Can revisit in that PR though

SimenB commented:
@thymikee I've reworked this to use feature detection instead of an option, and just dropped it from jsdom for now. It _should_ be invisible to consumers. Thoughts?

SimenB commented on packages/jest-environment/src/index.ts:
```diff
@@ -44,6 +44,11 @@ export declare class JestEnvironment {
   fakeTimersLolex: LolexFakeTimers | null;
   moduleMocker: jestMock.ModuleMocker | null;
   runScript<T = unknown>(script: Script): T | null;
+  compileFunction?<T = unknown>(
```
I've made it optional so we're not blocked by jsdom having support. It'll silently use it if it's ever implemented

SimenB commented on packages/jest-transform/src/enhanceUnexpectedTokenMessage.ts:
```diff
@@ -17,7 +17,8 @@ export default function handlePotentialSyntaxError(
   if (
-    e instanceof SyntaxError &&
+    // `instanceof` might come from the wrong context
+    e.name === 'SyntaxError' &&
```
the `SyntaxError` thrown by `compileFunction` comes from the passed in context

SimenB commented on packages/jest-environment-node/src/index.ts:
```diff
@@ -110,6 +116,20 @@ class NodeEnvironment implements JestEnvironment {
+}
+
+if (typeof compileFunction !== 'function') {
+  delete NodeEnvironment.prototype.compileFunction;
```
could potentially do some `supportsCompileFunction(): boolean` instead, but I think this is fine

thymikee replied:
Wouldn't it be more straight forward to make it a noop instead of deleting? Or there's some other advantage I don't see?

SimenB replied:
we need to use `runScript` if `compileFunction` is missing, how would we fall back to that? differentiating between `null` and `undefined` return seems flaky

thymikee replied:
Just checked back on the code in `jest-runtime` and you're right, this seems ok. Can you leave a comment on why we delete that and add a todo to remove it when we drop Node 8?

thymikee commented:
@SimenB works for me
